now ELAS library is used by default to compute horizontal disparity map from rectified images

GiuliaP opened the pull request:
- added [**ELAS**](http://www.cvlibs.net/software/libelas/) library's source code (`lib/elas_src`, `lib/elas_src_omp`).
- modified the `lib/CMakelists.txt` in order to compile ELAS (also `OpenMP` accelerated version on UNIX).
- added a class wrapper for ELAS (`elasWrapper.cpp`, `elasWrapper.h`) to deal with OpenCV's `cv::Mat`.
- added 2 functions to the `stereoCamera` class to initialize and dispose the ELAS wrapper.
- the `SFM` module initializes the ELAS wrapper by default: to use the previous OpenCV's SGMB you need to pass to the `SFM` module the `--use_sgbm` parameter.
- if the ELAS wrapper has been initialized, `stereoCamera` uses it to compute the disparity map (see `stereoCamera::computeDisparity` function).
- added 2 output ports to the `SFM` module to read the rectified images.

pattacini commented:
Thanks @GiuliaP 
could you provide doxygen documentation about the modifications you applied (e.g. the `--use_sgbm` option, the two new ports ...). I will be able then to recompile the documentation.

@Tobias-Fischer here's the PR; feel free to participate/test/contribute.

GiuliaP commented:
Sure, thank you!

Tobias-Fischer commented:
Hi @GiuliaP,

Thanks for the nice contribution! I quickly had a look at the source code, and except for some smaller comments it looks all good (see above & in https://github.com/GiuliaP/stereo-vision/commit/7b1eb3889396a2b2719dc93178be3a685d1fc05e, sorry for being picky ;) ).
I'll give it a try on our robot in the next couple of days.

Thanks again and best wishes,
Tobias

pattacini commented:
Thanks @GiuliaP :+1:

pattacini commented on lib/CMakeLists.txt:
```diff
@@ -27,17 +27,66 @@ if(USE_SIFT_GPU)
+  find_package(OpenMP)
+endif()
+
+if(OPENMP_FOUND)
```
Uhm, usually the `find_package(OpenMP)` directive should provide cmake variables like `OpenMP_*`, that are case-sensitive (e.g. see `OpenCV`). That might be working but I'll try the correct way as well. Be consistent also with the `message(STATUS OpenMP ...)` directives.

Tobias-Fischer replied:
Seems like in OpenMP they are not following that scheme for some reason: http://www.cmake.org/cmake/help/v3.0/module/FindOpenMP.html

pattacini replied:
gosh

pattacini commented on modules/SFM/SFM.h:
```diff
@@ -34,20 +34,23 @@ CopyPolicy: Released under the terms of the GNU GPL v2.0.
 parameters. For the stereo calibration see the module <a 
 href="http://wiki.icub.org/iCub/main/dox/html/group__icub__stereoCalib.html">stereoCalib</a>. 
-The module provides three output ports: the first one is the 
+The module provides five output ports: the first one is the 
```
@GiuliaP apparently, the doc for `--use_sgbm` option is still missing

GiuliaP replied:
@pattacini I'm sorry, I haven't updated the doc yet, I'm going to do it today. I've just pushed a commit fixing the issue on ELAS's directories.

pattacini replied:
I said doc but I meant the `SFM.h` file, just a line pointing out the new option is more than enough.

GiuliaP commented:
@Tobias-Fischer @pattacini Following your observations about the different settings of parameters, I'm modifying the code, providing the possibility of tuning some more parameters. I'll also do some other tests on our robot, trying to define the best configuration. I'll let you know when everything is ready to be tested!

pattacini commented:
yes, the idea is to test the final settings along with the `depth2kin` experiment, to see whether the finger will be always detected.

Tobias-Fischer commented:
Hi @GiuliaP: Thanks a lot for your effort! Please let me know once you tuned the parameter set on your robot, and I'll give it a try on ours.
Best, Tobias

Tariq-Abuhashim commented:
Hi;

Please note that unlike opencv, which performs dense texture matching, ELAS
estimates the dense disparity using MRFs (hence it is faster). Therefore,
these parameters are going to be scene/application dependent based on how
much smoothness is assumed.

Best regards;
Tariq Abuhashim

On Thu, Jun 11, 2015 at 10:38 AM, Tobias Fischer <notifications@github.com>
wrote:

> Hi @GiuliaP <https://github.com/GiuliaP>: Thanks a lot for your effort!
> Please let me know once you tuned the parameter set on your robot, and I'll
> give it a try on ours.
> Best, Tobias
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/robotology/stereo-vision/pull/1#issuecomment-111042954>
> .
>

GiuliaP commented:
@pattacini @Tariq-Abuhashim @Tobias-Fischer 
I have pushed another commit where I have:
- fixed a bug in the OpenMP accelerated version of LIBELAS (which I have reported to the authors)
- changed the way LIBELAS parameters can be set through the ResourceFinder of the `SFM` module, updating the related documentation in `SFM.h`
- now the default LBELAS setting is `ROBOTICS`, however the `ipol_gap_width` parameter is set to 40 as in the previous version
- unfortunately we have some troubles with the iCub here in Genoa therefore I haven't tested this version yet, @pattacini suggested me to push it anyway to see whether you can give it a try before us 
- therefore I haven't tested the `depth2kin` experiment yet (for which I suspect that it might be necessary to reduce the `ipol_gap_width` parameter, in order to detect the finger in a stable way)

pattacini commented:
Preliminary tests run with `depth2kin` along with default parameters of `SFM` showed a good segmentation!
